static/js/lib/auth.js

```diff
@@ -24,9 +28,78 @@ export const STATE_REGISTER_CONFIRM = "register/confirm"
 export const STATE_REGISTER_DETAILS = "register/details"
 export const STATE_REGISTER_EXTRA_DETAILS = "register/extra"
 
+export const ALL_STATES = [
+  STATE_ERROR,
+  STATE_ERROR_TEMPORARY,
+  STATE_SUCCESS,
+  STATE_INACTIVE,
+  STATE_INVALID_EMAIL,
+  STATE_USER_BLOCKED,
+  STATE_LOGIN_EMAIL,
+  STATE_LOGIN_PASSWORD,
+  STATE_LOGIN_PROVIDER,
+  STATE_REGISTER_EMAIL,
+  STATE_REGISTER_CONFIRM,
+  STATE_REGISTER_CONFIRM_SENT,
+  STATE_REGISTER_DETAILS,
+  STATE_REGISTER_EXTRA_DETAILS
+]
+
 export const generateLoginRedirectUrl = () => {
   const { pathname, search, hash } = window.location
 
   const next = `${pathname}${search}${hash}`
   return `${routes.login.begin}?${qs.stringify({ next })}`
 }
+
+export type StateHandlers = {
+  [AuthStates]: (response: AuthResponse) => void
+}
+
+const getErrorQs = (errors: Array<string>) =>
+  !isEmpty(errors)
+    ? qs.stringify({
+      error: errors[0]
+    })
+    : ""
+
+export const handleAuthResponse = (
+  history: RouterHistory,
+  response: AuthResponse,
+  handlers: StateHandlers
+) => {
+  /* eslint-disable camelcase */
+  const { state, redirect_url, partial_token, errors, field_errors } = response
+
+  // if this has a handler keyed off the state name invoke it
+  if (has(state, handlers)) {
+    handlers[state](response)
+  }
+
+  if (state === STATE_SUCCESS) {
+    window.location = redirect_url || routes.dashboard
+  } else if (state === STATE_LOGIN_PASSWORD) {
+    history.push(routes.login.password)
+  } else if (state === STATE_REGISTER_CONFIRM_SENT) {
+    history.push(routes.login.begin)
+  } else if (state === STATE_REGISTER_DETAILS) {
+    const params = qs.stringify({
+      partial_token
+    })
+    history.push(`${routes.register.details}?${params}`)
+  } else if (state === STATE_REGISTER_EXTRA_DETAILS) {
+    const params = qs.stringify({
+      partial_token
+    })
+    history.push(`${routes.register.extra}?${params}`)
+  } else if (state === STATE_USER_BLOCKED) {
+    history.push(`${routes.register.denied}?${getErrorQs(errors)}`)
+  } else if (
+    includes(state, [STATE_ERROR, STATE_ERROR_TEMPORARY]) &&
+    isEmpty(field_errors)
+  ) {
+    // otherwise we're in some kind of error state, explicit or otherwise
+    history.push(`${routes.register.error}?${getErrorQs(errors)}`)
```

Comment:
Need another if statement here to avoid being redirected to the registration errors page when entering an invalid email on the login page, something like `if(errors=["Couldn't find your account"]) {...do something}`